UM-USElab/gradecraft-development
# Remove Highest Points from GSE

### Status
READY

### Description
In addition to removing the _within one point_ validation on the GradeSchemeElements mass_edit page, we should try to remove/reduce our use of `highest_points` in its current state as much as possible.

By explicitly calculating the upper limits of a level, we aim to follow this refactor up by removing the `highest_points` column altogether on the GradeSchemeElement table. See `grade_scheme_element.rb` for revisions related to the calculation of ranges and upper-limits.

This PR also fixes a preexisting bug in master where the upper point limit for the highest GSE was being set incorrectly.

We may also want to consider renaming and/or adding a unique constraint to `lowest_points` and `course` in order to maintain data integrity, since there is currently only client-side validation stopping an instructor from creating multiple GSEs with the same point threshold for a given course.

### Related PRs
- One to rename column/add unique constraint to `lowest_points`, `course`, if desired
- One to remove `highest_points` column, when ready

### Todos

### Gem dependencies
- ADDED `binding_of_caller` for debugging, which was similarly done in #3034 

### Migrations
NO

### Steps to Test or Reproduce
- Ensure that editing and saving GradeSchemeElements works as expected
- Ensure that GSE-related views for students and instructors render as expected
- Verify that points to next level, progress percentage bars, and other statistics are correct

### Impacted Areas in Application
GradeSchemeElements

======================
Closes #3048 
Closes #3064
